src/Fmt_ast.ml

```diff
@@ -1186,10 +1186,7 @@ and fmt_expression c ?(box = true) ?epi ?eol ?parens ?ext
   let fmt_cmts = Cmts.fmt c.cmts ?eol pexp_loc in
   let fmt_atrs = fmt_attributes c ~pre:(fmt " ") ~key:"@" pexp_attributes in
   let parens = match parens with Some b -> b | None -> parenze_exp xexp in
-  let width xe =
-    String.length
-      (Format.asprintf "%t" (Cmts.preserve (fun () -> fmt_expression c xe)))
-  in
+  let width xe = String.length (Cmts.preserve (fmt_expression c) xe) in
```

Comment:
Here, the intent is to make it hard/impossible to mistakenly pass a fully-applied formatting function to preserve. @hhugo @gpetiot do you think this is enough?

Reply:
For my PR it is not enough.
If I replace `Location.width itm.pstr_loc <= c.conf.margin` with:
```
              let width i =
                let last = false in
                String.length (Cmts.preserve (fmt_structure_item ~last c) i)
              in
              width (sub_str ~ctx itm) <= c.conf.margin
```
I have a slightly worsed formatting:
```
let w =
  this one is toooooooooooooooooooooooooo
    (looooooooooooooooooooooooog but is (originally a one - liner))
```
is considered a one-liner and is packed with the preceding and following lines.
And I still have the "formatting changed ast" errors on `OCAMLFORMAT_STRUCTURE_ITEM_GROUPING=compact make -C test test`.